### ISSUE TITLE
InvalidCharError on range

### ISSUE THREAD

**Issue by adkabo** [JuliaLang/julia#44741]
I don't think this should throw.

```jl
julia> typemin(Char)
'\0': ASCII/Unicode U+0000 (category Cc: Other, control)

julia> typemax(Char)
'\xff\xff\xff\xff': Malformed UTF-8 (category Ma: Malformed, bad data)

julia> typemin(Char):typemax(Char)
ERROR: Base.InvalidCharError{Char}('\xff\xff\xff\xff')
Stacktrace:
 [1] throw_invalid_char(c::Char)
   @ Base ./char.jl:86
 [2] UInt32
   @ ./char.jl:133 [inlined]
 [3] codepoint
   @ ./char.jl:78 [inlined]
 [4] Type
   @ ./char.jl:51 [inlined]
 [5] -
   @ ./char.jl:226 [inlined]
 [6] (::Colon)(start::Char, stop::Char)
   @ Base ./range.jl:7
 [7] top-level scope
   @ REPL[6]:1
```

**Comment by StefanKarpinski** [comment]
This is debatable. I can see an argument that `typemax(Char)` should be `'\U0010ffff'` which is the largest valid Unicode character. But other times you want to uses `typemax(Char)` as an invalid `Char` that can be used as a sentinel value. You could write `'\0':'\U0010ffff'` instead.

To make this work we'd have to change `typemax(Char)` to return at least a well-formed character and maybe a valid character. The latter may be considered breaking however because I think there are places where `typemax(Char)` is used as an "impossible character" sentinel. If we wanted a well-formed but too-large-to-be-valid character value, that would be `"\xf7\xbf\xbf\xbf"[1]` aka `'\U1fffff'` (parser doesn't like this though). This actually already works (somewhat surprisingly):
```jl
julia> typemin(Char):"\xf7\xbf\xbf\xbf"[1]
'\0':1:'\U1fffff'
```
The big difference between the current `typemax(Char)` value and this, however, is that the sequence of bytes `[0x7f, 0xbf, 0xbf, 0xbf]` could appear in a string and would produce the invalid `Char` value `'\U1fffff'`. How is that different? Can't the sequence `[0xff, 0xff, 0xff, 0xff]`also occur? (Arguably much more likely as this is the encoding of `Int32(-1)`.) It can but that would be interated as four separate invalid `'\xff'` `Char` values. That's why the current `typemax(Char)` is a good sentinel: you can never get it from iterating or indexing into a string. Also, `c ≤ typemax(Char)` for all `c`.

**Comment by KristofferC** [comment]
> The latter may be considered breaking however because I think there are places where typemax(Char) is used as an "impossible character" sentinel.

One example: https://github.com/JuliaLang/Tokenize.jl/blob/0106feb8ce78a60983cb3e4a632970d6ab365303/src/utilities.jl#L31

**Comment by StefanKarpinski** [comment]
Yeah, it may have been a bad choice for `typemax(Char)`, but it's the choice that was made. The use of a larger integer value than you can actually get is inherited from C where it's quite common.

**Comment by MarkNahabedian** [comment]
I just stumpled on this issue folloring a link from the Julia Slack.

The documentation for `typemax` says

> The highest value representable by the given (real) numeric DataType.

Which clearly implies

```
typemax(Char) isa Char
true
```

Maybe that definition doesn't apply to `Char` becaose

```
Char <: Number
false
```

and, despite having 51 methods, `typemax` is on;y documented for numerics.

What does it mean for a Char to not be valid?  Does the sense in the error message match the definition of `isvalid`

```
isvalid(Char, typemax(Char))
false
```

Apparently so.

Begin rant:

The documentation for `typemax` should be broadened to descreibe how it should behave in general.  Perhaps it should have meaning for any type that has a meaningful total ordering.

End rant
